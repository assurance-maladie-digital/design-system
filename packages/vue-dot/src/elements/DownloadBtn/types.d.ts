export interface FileInfo {
	name: string;
	type: string;
}
