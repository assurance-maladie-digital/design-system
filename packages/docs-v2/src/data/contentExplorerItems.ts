import { TreeviewItem } from '~/components/content-explorer/types';

export const contentExplorerItems: TreeviewItem[] = [
	{
		name: 'public',
		path: 'public/index',
		children: [
			{
				name: 'css',
				path: 'public/css/index',
				children: [
					{
						name: 'base.css',
						path: 'public/css/base.css',
						file: 'css'
					},
					{
						name: 'loader.css',
						path: 'public/css/loader.css',
						file: 'css'
					}
				]
			},
			{
				name: 'js',
				path: 'public/js/index',
				children: [
					{
						name: 'lib',
						path: 'public/js/lib/index',
						children: [
							{
								name: 'matomo.js',
								path: 'public/js/lib/matomo.js',
								file: 'js'
							}
						]
					},
					{
						name: 'config.js',
						path: 'public/js/config.js',
						file: 'js'
					},
					{
						name: 'config.js.dist',
						path: 'public/js/config.js.dist',
						file: 'js'
					},
					{
						name: 'config.js.env',
						path: 'public/js/config.js.env',
						file: 'js'
					}
				]
			},
			{
				name: 'android-chrome-192x192.png',
				file: 'png'
			},
			{
				name: 'android-chrome-512x512.png',
				file: 'png'
			},
			{
				name: 'apple-touch-icon.png',
				file: 'png'
			},
			{
				name: 'favicon-16x16.png',
				file: 'png'
			},
			{
				name: 'favicon-32x32.png',
				file: 'png'
			},
			{
				name: 'favicon.ico',
				file: 'ico'
			},
			{
				name: 'index.html',
				file: 'html'
			},
			{
				name: 'robots.txt',
				file: 'txt'
			},
			{
				name: 'safari-pinned-tab.svg',
				file: 'svg'
			},
			{
				name: 'site.webmanifest',
				file: 'json'
			}
		]
	},
	{
		name: 'src',
		path: 'src',
		children: [
			{
				name: 'assets',
				children: [
					{
						name: 'logo.svg',
						file: 'svg'
					}
				]
			},
			{
				name: 'components',
				children: [
					{
						name: 'layout',
						children: [
							{
								name: 'tests',
								children: [
									{
										name: 'AppFooter.spec.ts',
										file: 'ts'
									},
									{
										name: 'AppHeader.spec.ts',
										file: 'ts'
									},
									{
										name: 'AppToolbar.spec.ts',
										file: 'ts'
									},
									{
										name: 'HeaderMenu.spec.ts',
										file: 'ts'
									}
								]
							},
							{
								name: 'AppFooter.vue',
								file: 'vue'
							},
							{
								name: 'AppHeader.vue',
								file: 'vue'
							},
							{
								name: 'AppToolbar.vue',
								file: 'vue'
							},
							{
								name: 'HeaderMenu.vue',
								file: 'vue'
							},
							{
								name: 'index.ts',
								file: 'ts'
							},
							{
								name: 'README.md',
								file: 'md'
							}
						]
					},
					{
						name: 'Links',
						children: [
							{
								name: 'tests',
								children: [
									{
										name: 'Links.spec.ts',
										file: 'ts'
									}
								]
							},
							{
								name: 'index.ts',
								file: 'ts'
							},
							{
								name: 'Links.vue',
								file: 'vue'
							}
						]
					}
				]
			},
			{
				name: 'constants',
				children: [
					{
						name: 'index.ts',
						file: 'ts'
					}
				]
			},
			{
				name: 'plugins',
				children: [
					{
						name: 'axios.ts',
						file: 'ts'
					},
					{
						name: 'form-builder.ts',
						file: 'ts'
					},
					{
						name: 'vue-dot.ts',
						file: 'ts'
					},
					{
						name: 'vuetify.ts',
						file: 'ts'
					},
					{
						name: 'webfontloader.ts',
						file: 'ts'
					}
				]
			},
			{
				name: 'router',
				children: [
					{
						name: 'index.ts',
						file: 'ts'
					},
					{
						name: 'navigationRedirect.ts',
						file: 'ts'
					},
					{
						name: 'routes.ts',
						file: 'ts'
					}
				]
			},
			{
				name: 'services',
				children: [
					{
						name: 'folders',
						children: [
							{
								name: 'api.ts',
								file: 'ts'
							}
						]
					},
					{
						name: 'dateFormatter.ts',
						file: 'ts'
					}
				]
			},
			{
				name: 'store',
				children: [
					{
						name: 'index.ts',
						file: 'ts'
					},
					{
						name: 'types.d.ts',
						file: 'ts'
					}
				]
			},
			{
				name: 'theme',
				children: [
					{
						name: 'colors.ts',
						file: 'ts'
					},
					{
						name: 'styles.scss',
						file: 'scss'
					}
				]
			},
			{
				name: 'translations',
				children: [
					{
						name: 'fr'
					},
					{
						name: 'index.ts',
						file: 'ts'
					}
				]
			},
			{
				name: 'types',
				children: [
					{
						name: 'index.d.ts',
						file: 'ts'
					}
				]
			},
			{
				name: 'views',
				children: [
					{
						name: 'tests',
						children: [
							{
								name: 'About.spec.ts',
								file: 'ts'
							},
							{
								name: 'Home.spec.ts',
								file: 'ts'
							},
							{
								name: 'Maintenance.spec.ts',
								file: 'ts'
							},
							{
								name: 'NotFound.spec.ts',
								file: 'ts'
							}
						]
					},
					{
						name: 'About.vue',
						file: 'vue'
					},
					{
						name: 'Home.vue',
						file: 'vue'
					},
					{
						name: 'Maintenance.vue',
						file: 'vue'
					},
					{
						name: 'NotFound.vue',
						file: 'vue'
					}
				]
			},
			{
				name: 'App.vue',
				file: 'vue'
			},
			{
				name: 'decorators.ts',
				file: 'ts'
			},
			{
				name: 'global.d.ts',
				file: 'ts'
			},
			{
				name: 'i18n.ts',
				file: 'ts'
			},
			{
				name: 'main.ts',
				file: 'ts'
			},
			{
				name: 'modules.d.ts',
				file: 'ts'
			}
		]
	},
	{
		name: 'tests',
		path: 'tests',
		children: [
			{
				name: 'e2e',
				children: [
					{
						name: 'plugins',
						children: [
							{
								name: 'index.js',
								file: 'js'
							}
						]
					},
					{
						name: 'specs',
						children: [
							{
								name: 'About.spec.ts',
								file: 'ts'
							},
							{
								name: 'Home.spec.ts',
								file: 'ts'
							},
							{
								name: 'NotFound.spec.ts',
								file: 'ts'
							}
						]
					},
					{
						name: 'support',
						children: [
							{
								name: 'commands.js',
								file: 'js'
							},
							{
								name: 'index.js',
								file: 'js'
							}
						]
					},
					{
						name: 'types',
						children: [
							{
								name: 'index.d.ts',
								file: 'ts'
							}
						]
					},
					{
						name: 'eslintrc.js',
						file: 'js'
					},
					{
						name: 'tsconfig.json',
						file: 'json'
					}
				]
			},
			{
				name: 'unit',
				children: [
					{
						name: 'App.spec.ts',
						file: 'ts'
					},
					{
						name: 'index.ts',
						file: 'ts'
					}
				]
			}
		]
	},
	{
		name: '.browserslistrc',
		path: '_browserslistrc',
		file: 'txt'
	},
	{
		name: '.editorconfig',
		path: '_editorconfig',
		file: 'txt'
	},
	{
		name: '.eslintrc.js',
		path: '_eslintrc.js',
		file: 'js'
	},
	{
		name: '.gitignore',
		path: '_gitignore',
		file: 'txt'
	},
	{
		name: 'babel.config.js',
		path: 'babel.config.js',
		file: 'js'
	},
	{
		name: 'cypress.json',
		path: 'cypress.json',
		file: 'json'
	},
	{
		name: 'jest.config.js',
		path: 'jest.config.js',
		file: 'js'
	},
	{
		name: 'package.json',
		path: 'package.json',
		file: 'json'
	},
	{
		name: 'postcss.config.js',
		path: 'postcss.config.js',
		file: 'js'
	},
	{
		name: 'README.md',
		path: 'README.md',
		file: 'md'
	},
	{
		name: 'tsconfig.json',
		path: 'tsconfig.json',
		file: 'json'
	},
	{
		name: 'vue.config.js',
		path: 'vue.config.js',
		file: 'js'
	},
	{
		name: 'yarn.lock',
		path: 'yarn.lock',
		file: 'txt'
	}
];
