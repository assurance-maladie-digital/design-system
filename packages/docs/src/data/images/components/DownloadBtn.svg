<svg width="541" height="357" viewBox="0 0 541 357" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="541" height="357" fill="#E7ECF5"/>
<rect x="63" y="88" width="415" height="181" fill="white"/>
<rect opacity="0.5" x="164" y="141" width="214" height="19" rx="9.5" fill="#CED9EB"/>
<g clip-path="url(#clip0_177_105652)">
<g clip-path="url(#clip1_177_105652)">
<path d="M236.667 200H234V196H230V200H227.333L232 204.667L236.667 200ZM227.333 206V207.334H236.667V206H227.333Z" fill="#0C419A"/>
</g>
<path d="M249.44 208V199.104H246.416V197.536H254.32V199.104H251.296V208H249.44ZM258.162 208.192C257.436 208.192 256.78 208.032 256.194 207.712C255.607 207.382 255.143 206.912 254.802 206.304C254.46 205.686 254.29 204.944 254.29 204.08C254.29 203.227 254.46 202.491 254.802 201.872C255.154 201.254 255.607 200.779 256.162 200.448C256.716 200.118 257.298 199.952 257.906 199.952C258.62 199.952 259.218 200.112 259.698 200.432C260.178 200.742 260.54 201.179 260.786 201.744C261.031 202.299 261.154 202.944 261.154 203.68C261.154 204.064 261.127 204.363 261.074 204.576H256.082C256.167 205.27 256.418 205.808 256.834 206.192C257.25 206.576 257.772 206.768 258.402 206.768C258.743 206.768 259.058 206.72 259.346 206.624C259.644 206.518 259.938 206.374 260.226 206.192L260.85 207.344C260.476 207.59 260.06 207.792 259.602 207.952C259.143 208.112 258.663 208.192 258.162 208.192ZM256.066 203.328H259.554C259.554 202.72 259.42 202.246 259.154 201.904C258.887 201.552 258.487 201.376 257.954 201.376C257.495 201.376 257.084 201.547 256.722 201.888C256.37 202.219 256.151 202.699 256.066 203.328ZM257.858 198.832L256.946 197.92L259.298 195.136L260.578 196.352L257.858 198.832ZM264.743 208.192C264.092 208.192 263.628 207.995 263.351 207.6C263.084 207.206 262.951 206.667 262.951 205.984V196.704H264.791V206.08C264.791 206.304 264.834 206.464 264.919 206.56C265.004 206.646 265.095 206.688 265.191 206.688C265.234 206.688 265.271 206.688 265.303 206.688C265.346 206.678 265.404 206.667 265.479 206.656L265.719 208.032C265.612 208.075 265.474 208.112 265.303 208.144C265.143 208.176 264.956 208.192 264.743 208.192ZM270.695 208.192C269.97 208.192 269.314 208.032 268.727 207.712C268.14 207.382 267.676 206.912 267.335 206.304C266.994 205.686 266.823 204.944 266.823 204.08C266.823 203.227 266.994 202.491 267.335 201.872C267.687 201.254 268.14 200.779 268.695 200.448C269.25 200.118 269.831 199.952 270.439 199.952C271.154 199.952 271.751 200.112 272.231 200.432C272.711 200.742 273.074 201.179 273.319 201.744C273.564 202.299 273.687 202.944 273.687 203.68C273.687 204.064 273.66 204.363 273.607 204.576H268.615C268.7 205.27 268.951 205.808 269.367 206.192C269.783 206.576 270.306 206.768 270.935 206.768C271.276 206.768 271.591 206.72 271.879 206.624C272.178 206.518 272.471 206.374 272.759 206.192L273.383 207.344C273.01 207.59 272.594 207.792 272.135 207.952C271.676 208.112 271.196 208.192 270.695 208.192ZM268.599 203.328H272.087C272.087 202.72 271.954 202.246 271.687 201.904C271.42 201.552 271.02 201.376 270.487 201.376C270.028 201.376 269.618 201.547 269.255 201.888C268.903 202.219 268.684 202.699 268.599 203.328ZM270.391 198.832L269.479 197.92L271.831 195.136L273.111 196.352L270.391 198.832ZM278.764 208.192C278.05 208.192 277.404 208.032 276.828 207.712C276.263 207.382 275.81 206.912 275.468 206.304C275.138 205.686 274.972 204.944 274.972 204.08C274.972 203.206 275.154 202.464 275.516 201.856C275.879 201.238 276.359 200.768 276.956 200.448C277.564 200.118 278.215 199.952 278.908 199.952C279.41 199.952 279.847 200.038 280.22 200.208C280.594 200.379 280.924 200.587 281.212 200.832L280.316 202.032C280.114 201.851 279.906 201.712 279.692 201.616C279.479 201.51 279.25 201.456 279.004 201.456C278.375 201.456 277.858 201.696 277.452 202.176C277.058 202.656 276.86 203.291 276.86 204.08C276.86 204.859 277.052 205.488 277.436 205.968C277.831 206.448 278.338 206.688 278.956 206.688C279.266 206.688 279.554 206.624 279.82 206.496C280.098 206.358 280.348 206.198 280.572 206.016L281.324 207.232C280.962 207.552 280.556 207.792 280.108 207.952C279.66 208.112 279.212 208.192 278.764 208.192ZM282.915 208V196.704H284.755V199.616L284.691 201.136C285.022 200.827 285.39 200.555 285.795 200.32C286.2 200.075 286.67 199.952 287.203 199.952C288.046 199.952 288.659 200.224 289.043 200.768C289.427 201.312 289.619 202.08 289.619 203.072V208H287.779V203.312C287.779 202.662 287.683 202.203 287.491 201.936C287.299 201.67 286.984 201.536 286.547 201.536C286.206 201.536 285.902 201.622 285.635 201.792C285.379 201.952 285.086 202.192 284.755 202.512V208H282.915ZM293.781 208.192C293.098 208.192 292.538 207.979 292.101 207.552C291.674 207.126 291.461 206.571 291.461 205.888C291.461 205.046 291.829 204.395 292.565 203.936C293.301 203.467 294.474 203.147 296.085 202.976C296.074 202.56 295.962 202.203 295.749 201.904C295.546 201.595 295.178 201.44 294.645 201.44C294.261 201.44 293.882 201.515 293.509 201.664C293.146 201.814 292.789 201.995 292.437 202.208L291.765 200.976C292.202 200.699 292.693 200.459 293.237 200.256C293.791 200.054 294.378 199.952 294.997 199.952C295.978 199.952 296.709 200.246 297.189 200.832C297.679 201.408 297.925 202.246 297.925 203.344V208H296.405L296.277 207.136H296.213C295.861 207.435 295.482 207.686 295.077 207.888C294.682 208.091 294.25 208.192 293.781 208.192ZM294.373 206.752C294.693 206.752 294.981 206.678 295.237 206.528C295.503 206.368 295.786 206.155 296.085 205.888V204.128C295.018 204.267 294.277 204.475 293.861 204.752C293.445 205.019 293.237 205.35 293.237 205.744C293.237 206.096 293.343 206.352 293.557 206.512C293.77 206.672 294.042 206.752 294.373 206.752ZM300.167 208V200.144H301.687L301.815 201.536H301.879C302.156 201.024 302.492 200.635 302.887 200.368C303.281 200.091 303.687 199.952 304.103 199.952C304.476 199.952 304.775 200.006 304.999 200.112L304.679 201.712C304.54 201.67 304.412 201.638 304.295 201.616C304.177 201.595 304.033 201.584 303.863 201.584C303.553 201.584 303.228 201.707 302.887 201.952C302.545 202.187 302.252 202.603 302.007 203.2V208H300.167ZM308.755 211.472C308.137 211.472 307.577 211.398 307.075 211.248C306.585 211.099 306.195 210.87 305.907 210.56C305.63 210.262 305.491 209.883 305.491 209.424C305.491 208.774 305.865 208.214 306.611 207.744V207.68C306.409 207.552 306.238 207.382 306.099 207.168C305.971 206.955 305.907 206.688 305.907 206.368C305.907 206.059 305.993 205.782 306.163 205.536C306.345 205.28 306.547 205.072 306.771 204.912V204.848C306.505 204.646 306.265 204.363 306.051 204C305.849 203.638 305.747 203.227 305.747 202.768C305.747 202.16 305.891 201.648 306.179 201.232C306.467 200.816 306.846 200.502 307.315 200.288C307.795 200.064 308.307 199.952 308.851 199.952C309.065 199.952 309.267 199.974 309.459 200.016C309.662 200.048 309.843 200.091 310.003 200.144H312.819V201.504H311.379C311.507 201.654 311.614 201.846 311.699 202.08C311.785 202.304 311.827 202.555 311.827 202.832C311.827 203.408 311.694 203.899 311.427 204.304C311.161 204.699 310.803 204.998 310.355 205.2C309.907 205.403 309.406 205.504 308.851 205.504C308.681 205.504 308.505 205.488 308.323 205.456C308.142 205.424 307.961 205.371 307.779 205.296C307.662 205.403 307.566 205.51 307.491 205.616C307.427 205.723 307.395 205.867 307.395 206.048C307.395 206.272 307.486 206.448 307.667 206.576C307.859 206.704 308.195 206.768 308.675 206.768H310.067C311.017 206.768 311.731 206.923 312.211 207.232C312.702 207.531 312.947 208.022 312.947 208.704C312.947 209.216 312.777 209.68 312.435 210.096C312.094 210.523 311.609 210.859 310.979 211.104C310.35 211.35 309.609 211.472 308.755 211.472ZM308.851 204.352C309.225 204.352 309.545 204.214 309.811 203.936C310.078 203.659 310.211 203.27 310.211 202.768C310.211 202.278 310.078 201.899 309.811 201.632C309.555 201.355 309.235 201.216 308.851 201.216C308.467 201.216 308.142 201.35 307.875 201.616C307.609 201.883 307.475 202.267 307.475 202.768C307.475 203.27 307.609 203.659 307.875 203.936C308.142 204.214 308.467 204.352 308.851 204.352ZM309.043 210.272C309.673 210.272 310.185 210.144 310.579 209.888C310.974 209.632 311.171 209.339 311.171 209.008C311.171 208.699 311.049 208.491 310.803 208.384C310.569 208.278 310.227 208.224 309.779 208.224H308.707C308.281 208.224 307.923 208.187 307.635 208.112C307.23 208.422 307.027 208.768 307.027 209.152C307.027 209.504 307.209 209.776 307.571 209.968C307.934 210.171 308.425 210.272 309.043 210.272ZM317.482 208.192C316.756 208.192 316.1 208.032 315.514 207.712C314.927 207.382 314.463 206.912 314.122 206.304C313.78 205.686 313.61 204.944 313.61 204.08C313.61 203.227 313.78 202.491 314.122 201.872C314.474 201.254 314.927 200.779 315.482 200.448C316.036 200.118 316.618 199.952 317.226 199.952C317.94 199.952 318.538 200.112 319.018 200.432C319.498 200.742 319.86 201.179 320.106 201.744C320.351 202.299 320.474 202.944 320.474 203.68C320.474 204.064 320.447 204.363 320.394 204.576H315.402C315.487 205.27 315.738 205.808 316.154 206.192C316.57 206.576 317.092 206.768 317.722 206.768C318.063 206.768 318.378 206.72 318.666 206.624C318.964 206.518 319.258 206.374 319.546 206.192L320.17 207.344C319.796 207.59 319.38 207.792 318.922 207.952C318.463 208.112 317.983 208.192 317.482 208.192ZM315.386 203.328H318.874C318.874 202.72 318.74 202.246 318.474 201.904C318.207 201.552 317.807 201.376 317.274 201.376C316.815 201.376 316.404 201.547 316.042 201.888C315.69 202.219 315.471 202.699 315.386 203.328ZM322.271 208V200.144H323.791L323.919 201.536H323.983C324.26 201.024 324.596 200.635 324.991 200.368C325.386 200.091 325.791 199.952 326.207 199.952C326.58 199.952 326.879 200.006 327.103 200.112L326.783 201.712C326.644 201.67 326.516 201.638 326.399 201.616C326.282 201.595 326.138 201.584 325.967 201.584C325.658 201.584 325.332 201.707 324.991 201.952C324.65 202.187 324.356 202.603 324.111 203.2V208H322.271Z" fill="#0C419A"/>
</g>
<rect x="208.5" y="184.5" width="135" height="35.0003" rx="3.5" stroke="#0C419A"/>
<defs>
<clipPath id="clip0_177_105652">
<rect x="208" y="184" width="136" height="36.0003" rx="4" fill="white"/>
</clipPath>
<clipPath id="clip1_177_105652">
<rect width="16" height="16" fill="white" transform="translate(224 194)"/>
</clipPath>
</defs>
</svg>
