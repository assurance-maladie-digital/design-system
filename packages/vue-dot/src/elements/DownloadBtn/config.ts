export const config = {
	menu: {
		offsetX: true,
		zIndex: 8,
		contentClass: 'vd-copy-tooltip-menu white--text text-body-2 ml-2'
	},
	btn: {
		icon: true
	}
};
