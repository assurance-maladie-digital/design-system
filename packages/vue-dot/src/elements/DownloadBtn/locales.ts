export const locales = {
	downloadSuccess: 'Document téléchargé avec succès.'
};
