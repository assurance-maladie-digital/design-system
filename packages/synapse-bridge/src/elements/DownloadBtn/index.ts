import DownloadBtn from './DownloadBtn.vue';

export default DownloadBtn;
