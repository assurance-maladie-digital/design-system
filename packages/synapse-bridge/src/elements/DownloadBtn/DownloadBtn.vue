<script lang="ts">
import { defineComponent } from "vue";
import type { PropType } from "vue";

import dayjs from "dayjs";
import deepMerge from "deepmerge";
import { mapActions } from "vuex";

import { AxiosResponse } from "axios";
import { parse } from "content-disposition-header";

import { mdiDownload } from "@mdi/js";

import { downloadFile } from "@/functions/downloadFile";

import { customizable, Options } from "@/mixins/customizable";

import { StateEnum } from "@/constants/enums/StateEnum";
import { IndexedObject } from "@/types";
import { ContentHeadersEnum } from "./ContentHeadersEnum";
import { FileInfo } from "./types";

import { config } from "./config";
import { locales } from "./locales";

export default defineComponent({
	inheritAttrs: false,
	mixins: [customizable(config)],
	emits: ["error"],
	props: {
		filePromise: {
			type: Function as PropType<() => Promise<AxiosResponse<Blob>>>,
			required: true,
		},
		fallbackFilename: {
			type: String,
			default: undefined,
		},
		notification: {
			type: [Boolean, String],
			default: locales.downloadSuccess,
		},
	},
	data() {
		return {
			locales,
			StateEnum,
			downloadIcon: mdiDownload,
			state: StateEnum.IDLE,
		};
	},
	computed: {
		btnOptions(): Options {
			return deepMerge<Options>(this.options.btn, this.$attrs);
		},
	},
	methods: {
		...mapActions('notification', ['addNotification']),
		getTimestampFilename(): string {
			return dayjs().format("YYYY-MM-DD - HH[h]mm[m]ss[s]");
		},
		getFileInfo(headers: IndexedObject): FileInfo {
			const contentType = headers[ContentHeadersEnum.TYPE];
			const contentDispositionHeader = headers[
				ContentHeadersEnum.DISPOSITION
			] as string;
			let filename: string | null = null;
			try {
				filename = parse(contentDispositionHeader).parameters.filename;
			} catch {
				filename = this.fallbackFilename || this.getTimestampFilename();
			}
			return {
				name: filename as string,
				type: contentType,
			};
		},
		notifyUser(): void {
			const message =
				typeof this.notification === "boolean"
					? locales.downloadSuccess
					: (this.notification as string);
			this.addNotification({
				type: "success",
				message,
			});
		},
		async download(): Promise<void> {
			this.state = StateEnum.PENDING;

			try {
				const { data, headers } = await this.filePromise();
				const { name, type } = this.getFileInfo(
					headers as IndexedObject
				);
				downloadFile(data, name, type);
				this.state = StateEnum.RESOLVED;
			} catch (error) {
				this.$emit("error", error);
				this.state = StateEnum.REJECTED;
				return;
			}
			if (this.notification) {
				this.notifyUser();
			}
		},
	},
});
</script>

<template>
	<VBtn
		v-bind="btnOptions"
		:loading="state === StateEnum.PENDING"
		class="vd-download-btn"
		data-testid="download-btn"
		@click="download"
	>
		<slot name="icon">
			<VIcon v-bind="options.icon">
				{{ downloadIcon }}
			</VIcon>
		</slot>

		<slot />
	</VBtn>
</template>

<style lang="scss" scoped>
.vd-download-btn :deep() {
	.v-btn__content {
		flex-wrap: wrap;
	}
	.v-icon {
		flex: none;
	}
}
</style>
