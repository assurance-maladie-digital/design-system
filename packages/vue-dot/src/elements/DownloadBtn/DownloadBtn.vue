<template>
	<VBtn
		v-bind="btnOptions"
		:loading="state === StateEnum.PENDING"
		class="vd-download-btn"
		@click.native="download"
	>
		<slot name="icon">
			<VIcon v-bind="options.icon">
				{{ downloadIcon }}
			</VIcon>
		</slot>

		<slot />
	</VBtn>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import dayjs from 'dayjs';
	import deepMerge from 'deepmerge';
	import { mapActions } from 'vuex';

	import { AxiosResponse, AxiosResponseHeaders } from 'axios';
	import { parse } from 'content-disposition-header';

	import { mdiDownload } from '@mdi/js';

	import { downloadFile } from '../../functions/downloadFile';

	import { customizable, Options } from '../../mixins/customizable';
	import { NotificationObj } from '../../modules/notification/types';

	import { StateEnum } from '../../constants/enums/StateEnum';
	import { ContentHeadersEnum } from './ContentHeadersEnum';
	import { FileInfo } from './types';

	import { config } from './config';
	import { locales } from './locales';

	const Props = Vue.extend({
		props: {
			filePromise: {
				type: Function as PropType<() => Promise<AxiosResponse<Blob>>>,
				required: true
			},
			fallbackFilename: {
				type: String,
				default: undefined
			},
			notification: {
				type: [Boolean, String],
				default: locales.downloadSuccess
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config));

	@Component({
		inheritAttrs: false,
		methods: mapActions('notification', ['addNotification'])
	})
	export default class DownloadBtn extends MixinsDeclaration {
		addNotification!: (obj: NotificationObj) => void;

		locales = locales;
		StateEnum = StateEnum;

		downloadIcon = mdiDownload;

		state = StateEnum.IDLE;

		get btnOptions(): Options {
			return deepMerge<Options>(this.options.btn, this.$attrs);
		}

		getTimestampFilename(): string {
			return dayjs().format('YYYY-MM-DD - HH[h]mm[m]ss[s]');
		}

		getFileInfo(headers: AxiosResponseHeaders): FileInfo {
			const contentType = headers[ContentHeadersEnum.TYPE];
			const contentDispositionHeader = headers[ContentHeadersEnum.DISPOSITION] as string;
			let filename: string | null = null;

			try {
				filename = parse(contentDispositionHeader).parameters.filename;
			} catch {
				filename = this.fallbackFilename || this.getTimestampFilename();
			}

			return {
				name: filename as string,
				type: contentType
			};
		}

		notifyUser(): void {
			const message = typeof this.notification === 'boolean' ? locales.downloadSuccess : this.notification as string;

			this.addNotification({
				type: 'success',
				message
			});
		}

		async download(): Promise<void> {
			this.state = StateEnum.PENDING;

			try {
				const { data, headers } = await this.filePromise();
				const { name, type } = this.getFileInfo(headers as AxiosResponseHeaders);

				downloadFile(data, name, type);

				this.state = StateEnum.RESOLVED;

				if (this.notification) {
					this.notifyUser();
				}
			} catch (error) {
				this.$emit('error', error);
				this.state = StateEnum.REJECTED;
			}
		}
	}
</script>

<style lang="scss" scoped>
	.vd-download-btn :deep() {
		.v-btn__content {
			flex-wrap: wrap;
		}

		.v-icon {
			flex: none;
		}
	}
</style>
