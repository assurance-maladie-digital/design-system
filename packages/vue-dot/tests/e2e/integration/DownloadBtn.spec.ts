import DownloadBtn from '@/elements/DownloadBtn';

import { filePromise } from '@/elements/DownloadBtn/tests/data/filePromise';

// Tests
describe('DownloadBtn', () => {
	it('renders correctly', () => {
		// Mount component
		cy.mountComponent(DownloadBtn, {
			propsData: {
				filePromise
			},
			slots: {
				default: 'justificatif.pdf'
			}
		}, true);
		cy.get('#__cy_root').toMatchSnapshot();
	});
});
