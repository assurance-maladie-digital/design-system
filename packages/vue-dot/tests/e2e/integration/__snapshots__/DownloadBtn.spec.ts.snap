exports[`DownloadBtn > renders correctly #0`] = `
<div id="__cy_root" class="v-application" data-app="true"></div>
`;
